src/client/Microsoft.Identity.Client/MsalServiceExceptionFactory.cs

```diff
@@ -26,7 +27,14 @@ internal class MsalServiceExceptionFactory
 
             if (IsInvalidGrant(oAuth2Response?.Error, oAuth2Response?.SubError) || IsInteractionRequired(oAuth2Response?.Error))
             {
-                ex = new MsalUiRequiredException(errorCode, errorMessage, innerException);
+                if (IsThrottled(oAuth2Response))
+                {
+                    ex = new MsalUiRequiredException(errorCode, MsalErrorMessage.AadThrottledError, innerException);

```

Comment:
"The request was throttled" is not an "UI required" kind of exception/error. `MsalUiRequiredException` is an exception that tells the app developer it needs to call a different API to show some interactive UI. This is an exception that the app commonly wants to catch under normal usage pattern. The throttled error on the other hand is an internal app error where the app sends requests too fast, generally due to wrong app design. The bug was asking to throw a different exception to discriminate between those two unrelated cases, as most apps will call the non-silent token retrieval when catching `MsalUiRequiredException`, which in the case of throttling not only doesn't work but make things worse.

Reply:
When a loop occurs (either due to app developer error or otherwise), the only way to break the loop is to ask user to login. So this exception is deliberately thrown. Also, in this case, the error message would be different "AadThrottledError" to distinguish it from the other exceptions.

Reply:
@SameerK-MSFT  although the case you describe _can_ happen, this is not the only case where throttling occurs.
- There's no loop in the throttling case I logged the original bug for; we simply login from several tests at once in parallel, and those tests are short-lived (sometimes a few seconds), so the shear amount of requests in a short period of time causes the throttling.
- Asking the user to login is unrelated, because throttling equally occurs with silent or UI-based login. The `MsalUiRequiredException` tells the app that some stored creds are not available for silent login, and therefore an interactive login needs to happen. This is very different from what throttling is about, which is the app making requests too fast. There should be a different exception to allow 1) during development the devs to catch this misuse (our case) and fix the app logic, and 2) during prod if the app cannot be fixed for any reason, implement an app-level throttling strategy to reduce the number of requests and avoid service-side throttling. With the current fix, the app needs to catch `MsalUiRequiredException`, then also branch on the content of the exception message (substring search), which I believe is bad practice, and a separate exception type should be used instead.